### PR TITLE
Support Checkmarx remote API

### PR DESCRIPTION
**Pull request by geoffwhittington**

### CODE REVIEW

**Comment by geoffwhittington** [comment]
@funkaoshi can you review?

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 1]
```diff
@@ -1,11 +1,153 @@
...
+        config.opts.add('analysis_method', 'Checkmarx server method', default='')
+        config.opts.add('analysis_server', 'Checkmarx server', default='')
+        config.opts.add('analysis_user', 'Checkmarx user name', default='')
+        config.opts.add('analysis_pass', 'Checkmarx password', default='')
+
+    def _get_api_url(self):
+        """
+        Retrieve the actual API URL we should use for integration
+        """
+        resolver = Client("%s://%s/Cxwebinterface/CxWsResolver.asmx?WSDL" % (self.config['analysis_method'], self.config['analysis_server']))
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Can wrap the long line:

```
url = "%s://%s/Cxwebinterface/CxWsResolver.asmx?WSDL" % (
    self.config['analysis_method'], self.config['analysis_server']
)
resolver  = Client(url)
```

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 2]
```diff
@@ -1,11 +1,153 @@
...
+        response = resolver.service.GetWebServiceUrl(APIVersion=1, ClientType="SDK")
+        if response.IsSuccesfull:
+            self.web_service_url = response.ServiceURL+"?WSDL"
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Any reason we don't do this in the constructor? It seems bad we have to make sure to call `_get_api_url` before using `self.web_service_url`. We could also make `web_service_url` a property:

@property
def web_service_url(self):
    if hasattr(self, '_web_service_url'):
        return self._web_service_url

    url = "%s://%s/Cxwebinterface/CxWsResolver.asmx?WSDL" % (
        self.config['analysis_method'], self.config['analysis_server']
    )
    resolver = Client(url)
    response = resolver.service.GetWebServiceUrl(APIVersion=1, ClientType="SDK")
    if not response.IsSuccesfull:
        raise CheckmarxIntegrationError("Failed to resolve API url: %s" % response.ErrorMessage)
    
    self.web_service_url = response.ServiceURL+"?WSDL"
```

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 3]
```diff
@@ -1,11 +1,153 @@
...
+            self.session_id = response.SessionId
+        else:
+            raise CheckmarxIntegrationError("Login unsuccessful: %s" % response.ErrorMessage)
+
+    def get_projects(self):
+        """
+        Retrieve the list of scan projects
+        """
+        response = self.checkmarx_wsclient.service.GetProjectsDisplayData(self.session_id)
+        if response.IsSuccesfull:
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Style. I like to write this out like:

```
if not response.IsSuccesfull:
    raise CheckmarxIntegrationError("Failed to retrieve list of projects: %s" % response.ErrorMessage)

return response.projectList
```

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 4]
```diff
@@ -14,17 +156,50 @@ class CheckmarxIntegrator(BaseIntegrator):
...
+            'network': 'https'
         }
+
+        config.opts.add('integration_mode', "Integration mode: (api or file)", default='file')
+        config.opts.add('project_name', 'Checkmarx Project name', default='')
+        config.opts.add('analysis_report_timeout', 'Maximum number of seconds to wait for report generation', default='240')
+
         super(CheckmarxIntegrator, self).__init__(config, self.TOOL_NAME, supported_input_types)
 
+    def initialize(self):
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Easier to read like this, I think. (I don't like calling super more than once in a subclassed method.)

```
def initialize(self):
    integration_mode = self.config['integration_mode'].lower()
    
    if integration_mode not in ['api', 'file']:
        raise UsageError("Invalid value for integration_mode. Valid values are: api or file")
        
    if integration_mode == 'api':
        for config_key in ['analysis_method', 'analysis_server', 'analysis_user', 'analysis_pass', 'project_name']:
            if not self.config[config_key]:
                raise UsageError("Missing value for option %s" % config_key)

        # disable file support
        self.supported_input_types.pop('file')
    
    super(CheckmarxIntegrator, self).initialize()

    try:
        self.config['analysis_report_timeout'] = int(self.config['analysis_report_timeout'])
    except (ValueError, TypeError):
        raise UsageError('Invalid value for analysis_report_timeout: %s' % self.config['analysis_report_timeout']
```

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 5]
```diff
@@ -14,17 +156,50 @@ class CheckmarxIntegrator(BaseIntegrator):
...
+            for config_key in ['analysis_method', 'analysis_server', 'analysis_user', 'analysis_pass', 'project_name']:
+                if not self.config[config_key]:
+                    raise UsageError("Missing value for option %s" % config_key)
+
+            # disable file support
+            self.supported_input_types.pop('file')
+            super(CheckmarxIntegrator, self).initialize()
+
+        elif self.config['integration_mode'].lower() == 'file':
+            super(CheckmarxIntegrator, self).initialize()
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Do we need to disable API support in this branch?

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 6]
```diff
@@ -32,11 +207,64 @@ def parse_report_file(self, report_file, report_type):
 
         self.findings = importer.findings
         self.report_id = importer.id
-
         return importer.findings, importer.id
 
+    def parse(self):
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: I'd drop the blank lines in this method.

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py [hunk 7]
```diff
@@ -32,11 +207,64 @@ def parse_report_file(self, report_file, report_type):
...
+        """
+        Retrieve the latest XML project report using the Checkmarx API
+        """
+        self.cx_api.login()
+        scan_id = self.retrieve_latest_scan_id()
+        report_id = self.cx_api.generate_report(scan_id)
+        report_status = self.cx_api.get_report_status(report_id)
+
+        # How long to wait until we check if the report is ready
+        seconds_until_retry = 4
```
Comment (funkaoshi) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: Make this a constant some place.

File: sdetools/modules/import_checkmarx/checkmarx_integrator.py (hunk 5 above)
Comment (geoffwhittington) [inline sdetools/modules/import_checkmarx/checkmarx_integrator.py]: I guess we could? But its not necessary. If `file` is defined as an option the parent class will complain that `report_file` is missing. There is nothing similar for `api`

**Comment by geoffwhittington** [comment]
@funkaoshi Code updated. I loved your feedback, thanks

**Comment by funkaoshi** [comment]
AH sorry I didn't look at this earlier. Let's merge!

**Comment by geoffwhittington** [comment]
Awesome I'll squash and merge. And make sure apiv2 tests pass

**Comment by funkaoshi** [comment]
:+1:

**Pull request merged**